Sources/BrowserServicesKit/Suggestions/SuggestionLoading.swift

```diff
@@ -21,141 +21,92 @@ import Foundation
-        var remoteSuggestions: [Suggestion]?
-        var remoteSuggestionsError: Error?
-        let group = DispatchGroup()
-
-        group.enter()
-        DispatchQueue.global().async {
-            // Bookmark suggestions
-            bookmarkSuggestions = self.bookmarkSuggestions(from: bookmarks, for: query)
-            group.leave()
-        }
+        let history = dataSource.history(for: self)
+        var apiResult: APIResult?
+        var apiError: Error?
 
-        // Remote suggestions
+        let group = DispatchGroup()
         group.enter()
         dataSource.suggestionLoading(self,
                                      suggestionDataFromUrl: Self.remoteSuggestionsUrl,
-                                     withParameters: [Self.searchParameter: query]) { [weak self] data, error in
+                                     withParameters: [Self.searchParameter: query]) { data, error in
             defer { group.leave() }
-            guard let self = self, let data = data else {
-                remoteSuggestionsError = error
+            guard let data = data else {
+                apiError = error
                 return
             }
-
-            do {
-                remoteSuggestions = try self.remoteSuggestions(from: data)
-            } catch {
-                remoteSuggestionsError = error
+            guard let result = try? JSONDecoder().decode(APIResult.self, from: data) else {
+                apiError = SuggestionLoaderError.parsingFailed
+                return
             }
+            apiResult = result
         }
 
-        group.notify(queue: .global()) {
-            let result = Self.result(maximum: maximum,
-                                     bookmarkSuggestions: bookmarkSuggestions,
-                                     remoteSuggestions: remoteSuggestions ?? [])
-
-            DispatchQueue.main.async {
-                guard !result.isEmpty || remoteSuggestionsError == nil else {
-                    completion(nil, SuggestionLoaderError.failedToObtainData)
-                    return
+        // 2) Processing it
+        group.notify(queue: .main) { [weak self] in
+            DispatchQueue.global(qos: .userInitiated).async {
+                let result = self?.processing.result(for: query,
+                                                     from: history,
+                                                     bookmarks: bookmarks,
+                                                     apiResult: apiResult)
+                DispatchQueue.main.async {
+                    if let result = result {
+                        completion(result, apiError)
+                    } else {
+                        completion(nil, SuggestionLoaderError.failedToProcessData)
+                    }
                 }
-                completion(result, nil)
             }
         }
     }
-
-    // MARK: - Bookmark Suggestions
-
-    static var minimumQueryLengthForBookmarkSuggestions = 2
```

Comment:
Whole processing of data into suggestions is wrapped into `SuggestionProcessing` class